### PR TITLE
New global pipelines for GoLang and Docker builds

### PR DESCRIPTION
**Pull request by ernestojeda**
This PR introduces two fully implemented Jenkins Declarative Pipeline to easily build docker images as well as build Golang based code.

The new pipelines are (_names are TBD_):
- [x] [edgeXBuildGoApp](vars/edgeXBuildGoApp.groovy)
- [x] [edgeXBuildDocker](vars/edgeXBuildDocker.groovy)
- [x] [edgeXGeneric](vars/edgeXGeneric.groovy)

I have created a few jobs on the Jenkins sandbox to demonstrate the new Jenkinsfile syntax.

https://jenkins.edgexfoundry.org/sandbox/view/All/job/geneva-app-functions-sdk-go-verify-pipeline/
https://jenkins.edgexfoundry.org/sandbox/view/All/job/geneva-app-service-configurable-verify-pipeline/
https://jenkins.edgexfoundry.org/sandbox/view/All/job/geneva-device-bacnet-c-verify-pipeline/
https://jenkins.edgexfoundry.org/sandbox/view/All/job/geneva-device-sdk-c-verify-pipeline/

More of the exact syntax will be written up in a future document.

Signed-off-by: Ernesto Ojeda <ernesto.ojeda@intel.com>

### CODE REVIEW

File: vars/edgeXSemver.groovy [hunk 1]
```diff
@@ -52,6 +52,16 @@ def call(command = null, credentials = 'edgex-jenkins-ssh', debug = true) {
...
         }
     }
+
+    env.setProperty('VERSION', semverVersion)
+
+    if(command == 'init') {
+        writeFile file: 'VERSION', text: semverVersion
+        stash name: 'semver', includes: '.semver/**,VERSION', useDefaultExcludes: false
+
+        sh 'docker run --rm ernestoojeda/artii -f slant Building version: v${VERSION}' // for fun
```
Comment (ernestojeda) [inline vars/edgeXSemver.groovy]: I need to remove this

File: vars/edgeXSetupEnvironment.groovy [hunk 2]
```diff
@@ -38,12 +40,6 @@ def call(vars) {
         }
     }
 
-    // attempty to set a default architecture
-    if(!env.ARCH) {
```
Comment (ernestojeda) [inline vars/edgeXSetupEnvironment.groovy]: This doesn't work the way really need to. The ARCH needs to be set on the VM where the build commands are being invoked. edgeXSetupEnvironment is too early in the process.

File: vars/edgex.groovy [hunk 3]
```diff
@@ -56,4 +56,87 @@ def didChange(expression, previous='origin/master') {
...
+
+    echo msg.join("\n")
+}
+
+def defaultTrue(value) {
+    (value == true || value == null) ? true : false
+}
+
+def defaultFalse(value) {
+    (value == false || value == null) ? false : true
```
Comment (ernestojeda) [inline vars/edgex.groovy]: New helper methods

File: vars/edgex.groovy [hunk 4]
```diff
@@ -56,4 +56,87 @@ def didChange(expression, previous='origin/master') {
...
+def mainNode(config) {
+    def defaultNode = config.nodes.find { it.isDefault == true }
+    defaultNode ? defaultNode.label : 'centos7-docker-4c-2g'
+}
+
+def nodeExists(config, arch) {
+    config.nodes.collect { it.arch == arch }.contains(true)
+}
+
+def getNode(config, arch) {
```
Comment (soda480) [inline vars/edgex.groovy]: I know we haven't talked about this, but we should start adding docstrings to all functions; providing a standard way of documenting functions.

File: vars/edgex.groovy [hunk 5]
```diff
@@ -56,4 +56,87 @@ def didChange(expression, previous='origin/master') {
...
+    ]
+
+    def _arch = config.arch ?: ['amd64', 'arm64']
+
+    println "Setting up nodes based on requested architectures [${_arch}]"
+
+    def _nodes = []
+
+    _arch.each { architecture ->
+        def n = defaultNodes.find { it.arch == architecture }
```
Comment (soda480) [inline vars/edgex.groovy]: I've never been a fan of one character variable names; I know they are acceptable to use for some constructs; iterating over indices (i) or key value pairs (k,v). maybe rename `n `to `foundNode`

Comment (ernestojeda) [reply vars/edgex.groovy]: thanks. I will incorporate

File: vars/edgex.groovy (hunk 4 above)
Comment (ernestojeda) [reply vars/edgex.groovy]: Will incorporate documentation at the next iteration

**Pull request merged**